AC-402: Provider Module - Fetch Providers

deepak140596 opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'AC-204 Applying MVP Model' -->
<!--- 'AC-JiraIssueNumber JiraIssueTitle' -->

## Description of what I changed
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->
- Added Provider Model
- Added Provider Repository and API Call
- Used LiveData to listen to Results in the Repository.
- Updated to Mockito 2, and Powermock to 1.7
- Added Android Arch core testing library to test AndroidX components
- Added Instant task executor
- Removed code for deprecated Hamcrest matchers and used Mockito's instead
- Ignoring unused stubs using Mockito.lenient. Unused stubs are very hard to find in a large project. Ignoring them ensures that atleast all our test cases pass.
- Replaced mockito.runner's MockitoJUnitRunner with mockito.junit's

Currently it only shows the available Provider list. **( Fetch )**

## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it -->
<!--- Just add the issue number at the end: -->
JIRA Issue: https://issues.openmrs.org/browse/AC-402

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).
<!--- No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one -->

- [ ] I have **added tests** to cover my changes. (If you refactored
existing code that was well tested you do not have to add tests)
<!--- No? -> write tests and add them to this commit `git add . && git commit --amend`-->

- [x] All new and existing **tests passed**.
<!--- No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works. -->

- [x] My pull request is **based on the latest changes** of the master branch.
<!--- No? Unsure? -> execute command `git pull --rebase upstream master` -->

deepak140596 commented:
![Capture+_2019-06-05-13-12-21](https://user-images.githubusercontent.com/34840023/58938963-a6ce7000-8793-11e9-89d0-51d17f07eb05.png)

deepak140596 commented:
@f4ww4z, I have used MVVM architecture the ViewModel will help in reusing the API calls from various other Activities/Fragments where the call is necessary. 

For example, I was working on Admission forms where it requires the knowledge of Providers (for input field "Admitted by"). Using ViewModels, we could easily get the Providers and then update the Admitted By filed there.

f4ww4z commented:
@deepak140596 Good start. Which of the [5 objectives ](https://issues.openmrs.org/browse/AC-402)did you complete here? Also see the merge conflict.

deepak140596 commented:
@f4ww4z , solved the merge conflicts.

Currently I am able to get all the Providers. From the [Provider REST API](https://wiki.openmrs.org/display/docs/REST+Web+Service+Resources+in+OpenMRS+1.9#RESTWebServiceResourcesinOpenMRS1.9-Provider) , I could only find Fetch, Create, Edit, Delete, Retire a Provider.

But in the JIRA issue, the description is of  Assign/Unassign patients to providers via relationship which can be found in [Provider Mangement Docs](https://wiki.openmrs.org/display/docs/Provider+Management+Module) with no link or description for any API to do that task.

Therefore, I am currently focussing on the CRUD operations on Providers and then move on to Creating the Dashboard for Assign/Unassign patients with more research on API.

f4ww4z commented:
Good work @deepak140596 . Sorry for the delay, I'm traveling, will be at home by tonight. Then I will test this thoroughly.

f4ww4z commented:
Also, please don't use a new design pattern. Follow the MVP to make the code consistent. This is more important. Just imagine other devs seeing more than 1 design pattern, it will make a lot of people confused. :smile:

deepak140596 commented:
@f4ww4z, Oh yes. It would really be a nightmare for future developers.
I changed it back to the MVP pattern.
Also I have used LiveData to listen to result in the Repository (all network calls will be done here). The Presenter will have no network calls. I think we should follow this for all Presenters. It will lead to a very clean code and architecture.

deepak140596 commented on openmrs-client/src/main/java/org/openmrs/mobile/databases/DBOpenHelper.java:
```diff
@@ -107,15 +107,16 @@ public long insertPatient(SQLiteDatabase db, Patient patient) {
 
```
This file has only been formatted by Android Studio. No change in contents.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/models/Provider.java:
```diff
@@ -0,0 +1,119 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Provider implements Serializable {
```
Will this model be saved in a database in the future? Then you can turn this into a Room entity.

f4ww4z replied:
As Provider is a resource, this class should extend `Resource`.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/models/Provider.java:
```diff
@@ -0,0 +1,119 @@
+    private String identifier;
+    @SerializedName("attributes")
+    @Expose
+    private List<Object> attributes = new ArrayList<Object>();
```
```suggestion
    private List<Object> attributes = new ArrayList<>();
```

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/models/Provider.java:
```diff
@@ -0,0 +1,119 @@
+    private Boolean retired;
+    @SerializedName("links")
+    @Expose
+    private List<Link> links = new ArrayList<Link>();
```
See above comment.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/providermanager/ProviderManagementRecyclerViewAdapter.java:
```diff
@@ -0,0 +1,74 @@
+    @Override
+    public void onBindViewHolder(ProviderManagementRecyclerViewAdapter.ProviderViewHolder holder, int position) {
+        final Provider provider = mItems.get(position);
+        if (provider.getPerson().getDisplay() != null)
```
See our [java conventions](https://wiki.openmrs.org/display/docs/Java+Conventions) on if statements.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/providermanager/ProviderManagementRecyclerViewAdapter.java:
```diff
@@ -0,0 +1,74 @@
+            holder.mIdentifier.setText(provider.getIdentifier());
+
+        holder.mRowLayout.setOnClickListener(v -> {
+            //Intent intent = new Intent(mContext.getActivity(), ProviderDashboardActivity.class);
```
https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotcommentoutcode

Instead, comment on what should be done in the future here.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/providermanager/ProviderManagementRecyclerViewAdapter.java:
```diff
@@ -0,0 +1,74 @@
+        public ProviderViewHolder(View itemView) {
+            super(itemView);
+            mRowLayout = (LinearLayout) itemView;
+            mIdentifier = (TextView) itemView.findViewById(R.id.providerManagementIdentifier);
```
Redundant casting.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/providermanager/ProviderManagementRecyclerViewAdapter.java:
```diff
@@ -0,0 +1,74 @@
+            super(itemView);
+            mRowLayout = (LinearLayout) itemView;
+            mIdentifier = (TextView) itemView.findViewById(R.id.providerManagementIdentifier);
+            mName = (TextView) itemView.findViewById(R.id.providerManagementName);
```
Redundant casting.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/api/ProviderRepository.java:
```diff
@@ -0,0 +1,59 @@
+package org.openmrs.mobile.api;
```
Add license header, and move this under `api/retrofit` package.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/api/ProviderRepository.java:
```diff
@@ -0,0 +1,59 @@
+    public LiveData<List<Provider>> getProviders(RestApi restApi){
+
+        final MutableLiveData<List<Provider>> data = new MutableLiveData<>();
+        if(NetworkUtils.isOnline()){
```
Format this file with AS default formatter.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/api/ProviderRepository.java:
```diff
@@ -0,0 +1,59 @@
+
+            restApi.getProviderList().enqueue(new Callback<Results<Provider>>() {
+                @Override
+                public void onResponse(Call<Results<Provider>> call, Response<Results<Provider>> response) {
```
Fix the warning `Not annotated parameter overrides @ParametersAreNonnullByDefault parameter`.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/api/ProviderRepository.java:
```diff
@@ -0,0 +1,59 @@
+                            data.setValue(null);
+                        }
+                    }else{
+                        Log.d(TAG,"Call Failed!");
```
Give a more descriptive error message.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/api/ProviderRepository.java:
```diff
@@ -0,0 +1,59 @@
+                        }
+                    }else{
+                        Log.d(TAG,"Call Failed!");
+                        ToastUtil.error("Call Failed!");
```
What call?

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/api/ProviderRepository.java:
```diff
@@ -0,0 +1,59 @@
+                @Override
+                public void onFailure(Call<Results<Provider>> call, Throwable t) {
+                    Log.d(TAG,"Call Failed!");
+                    ToastUtil.error("Call Failed!");
```
See above comments.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/api/ProviderRepository.java:
```diff
@@ -0,0 +1,59 @@
+
+    public LiveData<List<Provider>> getProviders(RestApi restApi){
+
+        final MutableLiveData<List<Provider>> data = new MutableLiveData<>();
```
Rename the variable to something more descriptive.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/api/ProviderRepository.java:
```diff
@@ -0,0 +1,59 @@
+    final String TAG = getClass().getName();
+
+
+    public LiveData<List<Provider>> getProviders(RestApi restApi){
```
Why did you choose to use `LiveData` and `MutableLiveData`, instead of RxJava? (just asking)

deepak140596 replied:
Currently there are no RxJava API calls in the app. RxJava is only called while accessing Database.
To use RxJava, we need to wrap the API calls in `Observable<>` instead of `Call<>` in the RetrofitApi interface which would create confusion as most API calls have been done in Presenters.

RxJava is pretty powerful than LiveData but to use it in our app currently, we need to make some big changes which should be tackled as another issue in JIRA. Also LiveData helps in easy wrapping of the API call results without much hassle while using the previous coding style.

I think that we will be using LiveData when we will port over to Room. So, a consistency will also be maintained.

Let me know your thoughts.

f4ww4z replied:
Yes, I agree, LiveData is recommended by Android Jetpack itself, and it's consistent with Room also. In a future JIRA issue, we can possibly replace the use of Promises in the other `Api` / `Repository` classes with LiveData too.

f4ww4z commented on openmrs-client/src/main/res/layout/fragment_provider_management.xml:
```diff
@@ -0,0 +1,39 @@
+  ~ Copyright (C) OpenMRS, LLC.  All Rights Reserved.
+  -->
+
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Consider using androidx's ConstraintLayout, for easier design.

f4ww4z commented on openmrs-client/src/main/res/layout/provider_management_row.xml:
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Use ConstraintsLayout instead.

f4ww4z commented on openmrs-client/src/main/res/values/strings.xml:
```diff
@@ -286,6 +286,11 @@
     <string name="charts_view_toolbar_title">Charts View</string>
 
+    <!--ProviderManager Activity-->
+    <string name="provider_manager">Provider manager</string>
```
```suggestion
    <string name="provider_manager">Manage Providers</string>
```

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/providermanager/ProviderManagementFragment.java:
```diff
@@ -0,0 +1,90 @@
+    @Override
+    public void updateAdapter(List<Provider> providerList) {
+        this.providerList = providerList;
+        mProviderManagementRecyclerView.getAdapter().notifyDataSetChanged();
```
This might cause null pointer. Consider using `Objects.requireNonNull` wrapper.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/providermanager/ProviderManagementFragment.java:
```diff
@@ -0,0 +1,90 @@
+    public void updateAdapter(List<Provider> providerList) {
+        this.providerList = providerList;
+        mProviderManagementRecyclerView.getAdapter().notifyDataSetChanged();
+        ProviderManagementRecyclerViewAdapter adapter = new ProviderManagementRecyclerViewAdapter(getContext(), providerList);
```
Why should it require to use a new adapter? The current adapter should already have been updated, as `notifyDataSetChanged()` was called on it in previous line.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/providermanager/ProviderManagementPresenter.java:
```diff
@@ -0,0 +1,60 @@
+
+public class ProviderManagementPresenter extends BasePresenter implements ProviderManagerContract.Presenter {
+
+    RestApi restApi;
```
This can be private.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/providermanager/ProviderManagementPresenter.java:
```diff
@@ -0,0 +1,60 @@
+        restApi = RestServiceBuilder.createService(RestApi.class);
+    }
+
+    public ProviderManagementPresenter(ProviderManagerContract.View view, RestApi restApi){
```
This constructor is never used. So add comments on why should this constructor stay, or else remove it.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/api/retrofit/ProviderRepository.java:
```diff
@@ -0,0 +1,76 @@
+
+            restApi.getProviderList().enqueue(new Callback<Results<Provider>>() {
+                @Override
+                public void onResponse(@NotNull Call<Results<Provider>> call, Response<Results<Provider>> response) {
```
Fix the warning `Not annotated parameter overrides @ParametersAreNonnullByDefault parameter` .

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/models/Provider.java:
```diff
@@ -0,0 +1,122 @@
+import java.util.List;
+
+@Entity(tableName = "provider_table")
+public class Provider extends Resource implements Serializable {
```
As `Resource` already implements `Serializable`, there's no need to implement again here.

f4ww4z commented on openmrs-client/src/main/res/layout/provider_management_row.xml:
```diff
@@ -0,0 +1,33 @@
+    android:layout_height="wrap_content"
+    android:background="@drawable/button_dialog_layout_background"
+    android:orientation="horizontal"
+    android:padding="16dp">
```
Use dimen's `activity_horizontal_margin`

deepak140596 commented:
@f4ww4z , I am having trouble creating test cases with LiveData in the Presenter. 
java.lang.NullPointerException
	at androidx.arch.core.executor.DefaultTaskExecutor.isMainThread(DefaultTaskExecutor.java:74)
	at androidx.arch.core.executor.ArchTaskExecutor.isMainThread(ArchTaskExecutor.java:116)
	at androidx.lifecycle.LiveData.assertMainThread(LiveData.java:442)

This error always pops up when calling the providerManagementPresenter.getProviders(fragment).

Any idea how to do testing with this?

f4ww4z commented:
@deepak140596 Did you see [this article](https://hackernoon.com/testing-livedata-on-android-d717c241efb1)? I don't have your code, so I cannot see what's causing the error.

deepak140596 commented:
`@PrepareForTest({NetworkUtils.class, ToastUtil.class, OpenMRS.class, OpenMRSLogger.class})
public class ProviderManagementPresenterTest extends ACUnitTestBase {

    //@Rule
    //public InstantTaskExecutorRule instantTaskExecutorRule = new InstantTaskExecutorRule();
    @Mock
    private RestApi restApi;
    @Mock
    private ProviderRepository providerRepository;
    @Mock
    private ProviderManagerContract.View providerManagerView;
    @Mock
    private Observer<List<Provider>> observer;
    @Mock
    MutableLiveData<List<Provider>> providerLiveData;

    List<Provider> providerList;

    private ProviderManagementPresenter providerManagementPresenter;
    private Fragment fragment = new Fragment();

    @Mock
    private OpenMRSLogger openMRSLogger;
    @Mock
    private OpenMRS openMRS;

    @Before
    public void setUp(){
        providerManagementPresenter = new ProviderManagementPresenter(providerManagerView,restApi);
        Provider providerOne = createProvider(1l,"doctor");
        Provider providerTwo = createProvider(2l,"nurse");
        providerList  = Arrays.asList(providerOne,providerTwo);
        providerLiveData.setValue(providerList);
        mockStaticMethods();
    }

    @Test
    public void shouldGetProviders_AllOK(){
        when(NetworkUtils.hasNetwork()).thenReturn(true);
        when(restApi.getProviderList()).thenReturn(mockSuccessCall(providerList));
        when(providerRepository.getProviders(restApi)).thenReturn(providerLiveData);

        providerRepository.getProviders(restApi).observeForever(observer);
        // this line causes problem
        providerManagementPresenter.getProviders(fragment);
       
     }
}`

deepak140596 commented:
I found the cause of the problem. We need to import `testImplementation "android.arch.core:core-testing:1.0.0"` to use the `InstantTaskExecutor` to execute onChange fuction immediately. Using this rule will (commented out line in the code above). 

But when I am importing it, all the Tests break.  This error pops up:
`Exception in thread "main" java.lang.NoClassDefFoundError: org/mockito/exceptions/Reporter
	at sun.reflect.GeneratedSerializationConstructorAccessor5.newInstance(Unknown Source)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.objenesis.instantiator.sun.SunReflectionFactoryInstantiator.newInstance`

It think it is due to the mockito, powermockito and the core-testing versions. I tried to change those to use Mockito 2 and Powermockito 1.7 with core-testing 2.0 (for AndroidX) but still there are many errors. It may be due to difference in coding style between Mockito 1 and 2 or something entirely different.

I have spent many hours researching this topic and still haven't found any solution.

deepak140596 commented:
@f4ww4z please address these issues. I went through your link. Some more links are
https://stackoverflow.com/questions/44270688/unit-testing-room-and-livedata

https://medium.com/@marco_cattaneo/unit-testing-with-mockito-on-kotlin-android-project-with-architecture-components-2059eb637912

https://stackoverflow.com/questions/45988310/setvalue-and-postvalue-on-mutablelivedata-in-unittest

https://stackoverflow.com/questions/48049131/cannot-resolve-symbol-instanttaskexecutorrule

https://stackoverflow.com/questions/48049131/cannot-resolve-symbol-instanttaskexecutorrule/56073388#56073388

f4ww4z commented:
@deepak140596 Did you try using the `AndroidJUnit4` runner?

deepak140596 commented:
@f4ww4z , I will test it and let you know asap.

deepak140596 commented:
@f4ww4z , the tests in ProviderManagementPresenterTest have successfully passed. Upgrading mockito, powermock and adding the android arch core testing library helped. Also some changes were made to DBOpenHelperTest where deprecated code was used. 

One problem arises though. Please see the Travis build failing. These are the same test cases that are failing on the my machine and now are reflected on the Travis build also. Some 13 test cases are failing from different Presenters.

deepak140596 commented:
@f4ww4z , all the test cases are now passing. 
Things I changed/Updated are listed below and also updated in the description.

1. Updated to Mockito 2, and Powermock to 1.7
2. Added Android Arch core testing library to test AndroidX components
3. Added Instant task executor
4. Removed code for deprecated Hamcrest matchers and used Mockito's instead
5. Ignoring unused stubs using Mockito.lenient. Unused stubs are very hard to find in a large project. Ignoring them ensures that atleast all our test cases pass.
6. Replaced mockito.runner's MockitoJUnitRunner with mockito.junit's

f4ww4z commented:
Great work @deepak140596 , will test it thoroughly once more. Do you think this will solve [AC-614](https://issues.openmrs.org/browse/AC-614)?

f4ww4z commented on openmrs-client/build.gradle:
```diff
@@ -120,7 +120,7 @@ dependencies {
             'androidx.test:rules:1.1.0-alpha4',
             'androidx.test:runner:1.1.0-alpha4',
-            'org.mockito:mockito-core:1.10.19',
+            'org.mockito:mockito-core:2.8.9+',
```
Did you see the 'Caution' at https://developer.android.com/studio/build/dependencies ?

f4ww4z commented on openmrs-client/build.gradle:
```diff
@@ -159,11 +159,12 @@ dependencies {
     implementation 'com.hbb20:ccp:2.2.7'
     testImplementation(
-            'org.mockito:mockito-core:1.10.19',
+            'org.mockito:mockito-core:2.8.9+',
```
Did you see the 'Caution' at https://developer.android.com/studio/build/dependencies ?

deepak140596 commented:
@f4ww4z , AC-614 is an entirely different issue. I tried solving that but couldn't. I will keep on researching on that and once the solution is found, I will update.
